packages/unplugin-typegpu/src/index.ts

```diff
@@ -14,10 +14,51 @@ import {
 } from './common';
 
 type TgslFunctionDef = {
-  varDecl: CallExpression;
-  implementation: AnyNode;
+  implementation: acorn.AnyNode;
+  functionDeclarationName?: string | undefined;
 };
 
+function isKernelMarkedFunction(
+  node:
+    | acorn.FunctionDeclaration
+    | acorn.AnonymousFunctionDeclaration
+    | acorn.FunctionExpression
+    | acorn.ArrowFunctionExpression,
+) {
+  if (node.body.type === 'BlockStatement') {
+    for (const statement of node.body.body) {
+      if (
+        statement.type === 'ExpressionStatement' &&
+        statement.directive === 'kernel'
+      ) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function removeKernelDirectiveFromFunction(
+  node:
+    | acorn.FunctionDeclaration
+    | acorn.AnonymousFunctionDeclaration
+    | acorn.FunctionExpression
+    | acorn.ArrowFunctionExpression,
+) {
+  if (node.body.type === 'BlockStatement') {
+    node.body.body = node.body.body.filter(
+      (statement) =>
+        !(
+          statement.type === 'ExpressionStatement' &&
+          statement.directive === 'kernel'
```

Comment:
babel removes directive expressions (keeps them on node.body.directives). maybe we should remove all of them in rollup to keep tinyest from diverging between the two parsers?